### DISCUSSION TITLE
What is the future Microsoft?

### DISCUSSION THREAD

**Discussion by wstaelens** [dotnet/maui#4879] [General]
// Start rant.

In the past it was very simple:
Desktop: **WinForms**
Web: **ASP.NET**

Next we had:
Desktop: WinForms, **WPF**
Web: ASP.NET, **Silverlight**

It was all simple, Microsoft and we as developers had a clear view on the future. If we invested in new development/projects as an ISV, we know it would be supported etc...

Now it is all worse and shit.
As a developer now, we don't know what to choose, what the future of Microsoft would be. 

Current situation: we have existing WinForms and WPF projects. We want to modernize them, if possible make them in a first stage available for Windows only but keep the options open for other platforms (Android, Mac, ...).

What to choose? 
WinForms, WPF, UWP, MAUI .NET, WinUI 3, .NET MAUI with Blazor, Xamarin, Project reunion, ...?

We are porting libraries to .NET Standard / .NET 5 / 6 / ... meaning we can't use them in WinForms / WPF anymore at this moment.

If I take a look at .NET MAUI it is lacking basic controls (tabcontrol, menu). 
Should we go to WinUI 3? (is it the same as UWP? Will it be replaced by MAUI?)

It is really frustrating.
As an ISV, where should you invest in? In the past we've invested in Silverlight, "UWP", Windows Phone and all went unsupported.
What should we use now? What will be supported? Or what doesn't have a long-term future? 
I feel like WinForms and WPF are still the most mature Windows technologies, but they are not giving us a good future-proof feeling anymore.

(For the web it is clear. ASP.NET (Core) 5 / 6 / 7 and Blazor.)

// End rant

**An existing WPF application should be ported to ... ?**

**Comment by saint4eva** [comment]
Nonsense rant.

**Comment by GeraudFabien** [comment]
```
We are porting libraries to .NET Standard / .NET 5 / 6 / ... meaning we can't use them in WinForms / WPF anymore at this moment.
```
Why ?
- WPF/Winform .Net framework can be used with .NetStandard2.0 compatible app.
- WPF/Winform .Net core can be used with .NetStandard2.0 or .Net core > 3.1

```
We want to modernize them, if possible make them in a first stage available for Windows only but keep the options open for other platforms (Android, Mac, ...).
```
If you want to access android, mac, ... you can use Maui or Maui Blazor. You can also use Blazor on WPF to do the migration.

```
WinForms, WPF, UWP, MAUI .NET, WinUI 3, .NET MAUI with Blazor, Xamarin, Project reunion, ...?
```
- WinForms : Legacy
- WPF : Windows only (Windows 7)
- UWP : XBox, Hololens, ... You can now migrate to WinUI2 without too much problem. Then try to move to WinUI3.
- MAUI .NET : It's call MAUI. It's only for native UI. The framework force you to redo you're UI on all plateform. If you want to do Native and cross plateform go for it. But you may look for UNO or Blazor otherwise.
- WinUI 3 : Next-gen windows native ui. You can use it through other framework (https://microsoft.github.io/react-native-windows/docs/winui3). Maui use it too. Any native UI use it. It's evolve from WinUI2.0 but it's not compatible
- .NET MAUI with Blazor :  It's only blazor on MAUI. It allow to do real cross plateform UI. (With the same UI in all plateform)
- Xamarin : Old name of Maui. Not retro compatible with Maui
- Project reunion : I don't understand what the use in this list. It's not a UI framework. 

```
As an ISV, where should you invest in? In the past we've invested in Silverlight, "UWP", Windows Phone and all went unsupported.
```
As a pro you have to thinks before migrating. Why do you migrate. Is one of you're client specifically asking for a specific feature (Or platform) you don't have?
Also never ever use a new framework right after the first release. Whatever the idea on internet.
- Personal project : 0 year after release 1 (It's not professional)
- POC : 1 year after release 1 (I start to trust blazor enough for poc since .Net6.0 release)
- Really small project (less than 6 month and for one dev) : 3 year after the first release.
- New project : 3-5 years (Depending on how much i need it).
- Migrate : IDK but at least 5 years. You need something stable.

Using a framework need to be thinks. It's okay to want the new shiny framework. But has a pro you can't use some thinks new without knowing if it is stable enough. Also waiting would increase the chance to found a way to fix any blocking problems you may encounter directly on internet (Stackoverflow). 

```
It is really frustrating.
```

Maybe you should learn of you're mistake. .Net doesn't have a lot of framework. What did you do with javascript? New project pop and died every day. Do you use them all ? It's not pokemon it's you're job.

```
Should we go to WinUI 3?
```
If you need it do the migration. If you don't .... well don't. Too young for me. Every thinks you said make me thinks of all the people using beta and complaining after about bug, doc or tooling. Well if you want to use a beta or a new tool and have problem. Well it's you're mistake and maybe you should avoid using new thinks day 1.

**Reply by wstaelens** [reply]
For example: .NET Core 1, 2, 3, ... .NET 5, .NET 6 is lacking good XPS and print support and has issues with the compression library, a part was/is not implemented and was left out.
Microsofts own print drivers generate files that .NET cannot read because it is not supported/implemented, (works fine in .NET Framework though). Not sure it will even be fixed in .NET 7...
This and other things are blocking us from fully moving to .NET. (yes some libraries can be ported to .NET Standard / .NET) but not everything. So we are for some projects stuck with "legacy" winforms on .NET Framework.

Again it is half half work being released. Some tickets are already out there for several years and have not been fixed.

> Project reunion : I don't understand what the use in this list. It's not a UI framework.

--> I was referincing to this older doc: https://docs.microsoft.com/en-US/windows/apps/winui/winui3/release-notes/release-notes-08-preview

@saint4eva calls it nonsense rant, I disagree.  When you dive into document conversion, rendering, printing, fonts, etc... you'll notice lots of problems when trying to "modernize" your apps.

You can close this rant, but I hope some read it. 
Too many unfinished/preview stuff and a lot of projects that have no or maybe a short life span. + not every library is yet at .NET 5 and it is already long-term support deprecated and we should move already to .NET 6, while external vendors still need to update their packages to .NET 6.

(and blazor, we are working on this, but even DevExpress seems to struggle providing production ready controls for months/weeks now.)

**Reply by GeraudFabien** [reply]
To me Project reunion = WinUI3. Even through **if my memory work well** it's more of a bridge to bring WPF/WinForm control to WinUI3. So it's not a ui framework.

WPF ==> UWP ==> WinUI2 ==> WinUI3
UWP didn't died it became WinUI2. It's not library compatible. But it should be mostly code compatible. So you should not have a lot of work to move from one to the other.
WinUI2 change to became WinUI3. There are some change in namespace that make the migration more difficult.

Again has i say framework and virus have one things in common. There not like pokemon and you don't have to catch them all. Wait two year that WinUI3 mature. Use it in poc or personal project (Or encourage you're dev to use it for they're personal project). And in 2024-2025 if it seems to work well you may use it in a real project.

To resume 
- If you're app is in WPF and you don't need to change. Don't change. You may make some **new** control in WinUI2 and integrate them in you're WPF. But do not migrate the legacy. 
- If you're app is in UWP and you don't need to change. You may move some part in WinUI2 but you can stay on UWP.
- If you're app is in WinUI2 and you don't need to change. Don't change. Try regularly 
- If you're app is in WinUI3 and you don't need to change. you don't need to go to MAUI.

But what do i mean by you don't need to change. I mean one of you're client ask for a feature you can't do today.
- On WPF you can't target XBOX, Hololens, ...
- On UWP/WinUI2 some feature where never develop.
- On WinUI3.0 you can't target mobile platform.

Again you may also want to look at [UNO](https://platform.uno/) app. https://platform.uno/docs/articles/uno-development/working-with-uno-wpf-host.html
(PS: I have a lot of hope about UNO and blazor. But like i say before i would never use them in a real project until 2025 for Uno and 2024 for blazor).

https://github.com/dotnet/wpf/issues/3546 I never used XPS or needed it in my life.

```
issues with the compression library
```
I can't find this thread on your activity but i use this https://github.com/icsharpcode/SharpZipLib personally. It's great. You can do everything in memory. Witch saved me in the past.

https://github.com/microsoft/microsoft-ui-xaml/issues/4650 You have already ask. Why do you open a new thread here ?

**Reply by wstaelens** [reply]
seems like I forgot there was a thread already, sorry.

regarding XPS: https://github.com/dotnet/wpf/issues/3546#issuecomment-825061189

**Comment by yyliveyy** [comment]
Microsoft has been through chaos, but now it's also moving toward unity. The new project adopts MAUI, I believe it will bring long-term support after going through the chaos. We should do more maintenance work for old projects.

**Reply by janseris** [reply]
still waiting for MAUI being not broken

**Reply by Transis-Felipe** [reply]
Me too

**Comment by Xyncgas** [comment]
https://github.com/dotnet/aspnetcore/issues/38745

**Comment by akhanalcs** [comment]
Here's my take on this:
1. For native windows app development, go with Win UI 3. Don't even worry about WPF, UWP, Project Reunion etc.
Watch [this video](https://www.youtube.com/watch?v=1ZBMHFoMVAs) for more info.
If you need to target more than windows, just go with .NET MAUI. (MAUI uses WinUI 3 library under the covers for windows apps).

2. For cross platform native app development, go with .NET MAUI. Note: .NET MAUI doesn't support Linux yet.
 
3. For Web apps, go with Blazor (of if you prefer JS, use .NET core backend and React/ Angular frontend).
 
4. For Pixel perfect apps across all platforms (including Linux), go with [Avalonia](https://avaloniaui.net/) (Better than Uno and Flutter IMO).

**Conclusion**: I'd go with Avalonia to target **everything** and call it a day. It's amazing!

**Comment by wstaelens** [comment]
update: currently doing Balzor (Server) and waiting for native windows app development as WPF is still more powerfull and has more decent controls compared to WinUI 3. Don't know if I will go with WPF or WinUI3 (or MAUI) for Windows desktop apps.
(experimenting with .NET MAUI but still early)

**Reply by tharlab** [reply]
msft give us alotof option, and you are confusing on it,
so dont blame msft blame ur self, cz you cant choose right technology for your need,
and also you said above 
>`(For the web it is clear. ASP.NET (Core) 5 / 6 / 7 and Blazor.)`

blazor is mature since 2016 befor flutter born and day by day it more2 stronger and popular,
and if you know about "blazor wasm maui" it will be impressed you,
not only running on web like ur statemnt but also on every ecosystem every where

**Reply by GeraudFabien** [reply]
> blazor is mature since 2016

Blazor first stable release at the end of 2020 (middle for blazor wasm). Blazor didn't even have CSS isolation at that time. And rewriting header was a headache... Blazor is good but it need time to be mature. Net 6.0 blazor is the first version i may consider "mature".  But it's too soon to use in a real project safely.

**Comment by Symbai** [comment]
WinUI 3 is just as buggy as MAUI. Have you seen the long existing issue of setting a window title crashes the application? It hasn't been fixed and is just one of the great examples on how the easiest things don't work. Winforms and WPF are at least stable and that's what is most important for bigger applications.

Maui is also very buggy. I started with Maui few weeks ago and discovered so many bugs (you can search for it) and some of the bugs are major issues that can block you until they are fixed and a new Maui update is released with that fix included. Even the tools for MAUI are often broken. Means you cannot debug on iOS or you cannot publish using a mac host. Hot reload doesnt work half of the time and whole VS crashes if you add a new event in XAML while debugging just to name a few.

The worst decision Microsoft has made was to take WPF away from the original authors and assigned a different development team to it. Now its completely stuck in extremely basic maintenance mode with absolutely no hope for any progress. It is that worse because WinUI3 is clearly NOT ready, not even considerable, not even in a beta stage. There is NO serious alternative to this popular framework for windows desktop.

 If Microsoft isn't hiring more developers and improve support and development speed, in my opinion other UI frameworks are probably the better choice if you want to be ready for the future. Winforms and WPF future is unknown. WinUI3 and Maui might leave you stuck or unsuccessful to achieve what you want. And both aren't really something new, they are just renamed and re-target code bases from UWP and Xamarin. How long can you wait until the handful of controls become stable? Another 3 or 5 years?

**Reply by wstaelens** [reply]
Same feeling @Symbai , and even on WPF/XPS the lack of bug fixes is making me worry even about this stable product.
Our customers can not install any alpha,preview,beta WinUI3/MAUI software in production.

**Reply by janseris** [reply]
@Symbai the truth has been spoken.

**Comment by janseris** [comment]
WPF/WinForms - migrate to .NET 6 if you need to and everything will be fine
Blazor = web (or Razor pages for static content)
MAUI = broken but possible to use for Android

**Reply by manishbisen** [reply]
In above post I saw peoples saying that MAUI is buggy.Can any one from you tell me name of any product which is without bugs.Nothing is perfect in this world.

**Reply by wstaelens** [reply]
@manishbisen maybe "not production/enterprise ready in the next x-months/years" might be a better wording instead of buggy".

**Reply by janseris** [reply]
@wstaelens well the worst thing is that the bugs are not being solved. And MAUI is taking bugs from WinUI and Xamarin into itself while adding own bugs as well.
 Today I read that CollectionView and ListView performance (more than 100 items) is slow because it is slow in Xamarin and as been for many years and reported and nothing has been fixed. 
 Same for WinUI MediaPicker returning null for everything. It has been already known but nothing was done about it and it is used in MAUI.

**Comment by ccarlo88** [comment]
All this mess was started by Nadella when he took office in 2014. Before him, there was a planning, logical roadmap with evolution AND with things that made sense.
MAUI was dead before started. No one used Xamarin in the past and it was already known as being bad, REALLY bad. 
Typical from current administration, get something which is already bad and no one uses (in this case Xamarin) and insist on it, rename it, don't care about results and keep insisting on a bad decision and smiling pretending a good job is being done. 

Microsoft lacks the seriousness it once had and now everything is just a party............. oh, what a coincidence, there are some other competitors out there acting in the exact same way for years also without much success. 🤣

**Reply by manishbisen** [reply]
@ccarlo88 It seems like you got some money from Google or Meta to put such kind of comments here 🤔

**Reply by akhanalcs** [reply]
@manishbisen 
Hello Manish, Truth hurts but @ccarlo88 is speaking the truth here. 
Satya Nadella is the kind of guy who [has no real vision](https://www.reddit.com/r/dotnet/comments/v5petk/microsoft_is_lagging_behind_in_innovation_these/?utm_source=share&utm_medium=web2x&context=3) or direction and does everything 'half-assed'. Look at the guy, he can't even give Windows on ARM a serious consideration when the whole industry is moving to ARM for many benefits it brings. 
If Microsoft has enough money to buy 60 billion + dollars game studios, it should have enough money to properly invest on MAUI, Windows on ARM, Cloud technologies and so on. 
There are 1.6 k active bugs in MAUI and every new bug is pushed to the backlog. Sounds like they have 5 developers who work on MAUI and expect everything from the community, it's just ridiculous and recipe for failure. 
All this rant is coming from a developer who loves C# and .NET and wants to see them succeed. 

Satya is not the right CEO for Microsoft IMHO. Someone like Sundar Pichai can do wonders at Microsoft, but not Satya. 🙅‍♂️

**Reply by ccarlo88** [reply]
@manishbisen I don't receive money from any of the "giant" corps and in the end, I make my comment saying that "competitors do the same and go nowhere". So, it is self-implied that I don't work for Google/Meta or any other of them
@affableashish Ballmer had a solution where people would program for Windows and Window Phone on the same solution and share some things. The natural process, would have been integrating Android  to that solution but instead the past was ignored and just started with a solution no one was using (Xamarin) instead of something which was successful. If I were the responsible I would discard Xamarin/MAUI and would do something on top of what people really use and which is known to work. I don't know a lot of Pichai but he also didn't deliver anything really meaningful so far (at least as far as I know).

**Reply by manishbisen** [reply]
I agree that day by day number of bugs are increasing  in .net maui which is really point of concern. But another fact is that if you look at to Flutter currently it is having 5k+ open issues but still many developers are loving it.I think Maui is new born so it will take some time.

**Reply by ccarlo88** [reply]
@manishbisen I'm not talking about Flutter and I don't even want to enter into the Google subject here. I'm talking about a bad decision which was made and, instead of stopping with it to do a better alternative, it continues endlessly. In 10 years people will still have the same discussion about the same things, and some people will try to say: "yeah my grass is not that bad, the grass of the neighbor 5 blocks away is worst than mine". I mean, this same conversation happened 3-4 years ago and now I see people pissed even about other subjects on other repos.

**Reply by janseris** [reply]
@manishbisen if you make a graph of maui bug issues in time and try the technology, you will see that it is turning very unmaintainable. People will not love it because it will never be fixed in the current tempo.
And a fix for that is more developers and more testers.

**Comment by saint4eva** [comment]
https://dotnet.microsoft.com/en-us/platform/customers/xamarin

**Reply by ccarlo88** [reply]
Maybe they should also include "Hello world" to account for the statistics... 🤣 
Funny, they should also include WPF/WinUI numbers there, maybe they would start realizing in which year things started going downhill....

**Reply by saint4eva** [reply]
It seems you do not have anything meaningful to say?

**Comment by manishbisen** [comment]
@ccarlo88 Microsoft doesn't matter what you , me or anybodies else think about it. The fact of matter 
is that Big companies do what it's best for the business based on their analysis, planning, study,feedback etc.
BTW you can check below link which contains list of snppt of maui & xamarin & I bet you it doesn't contain hello world.
https://www.snppts.dev/

**Reply by ccarlo88** [reply]
Which I don't care if they do or do not but now we should start reviewing when open source is really a non abiding job.
- Big corp: "Let's do the open source world better and go to this way."
- Contributor: "Yeah, I contribute this xxx... could we put the direction to a vote?"
- Big corp never answer: "......."

As a big company is driving the directions, I see that open source is nothing more than an employment without payment, people are conditioned to do what the company wants.
I believe it is the time law take part into this chapter in history.
